### ISSUE TITLE
patch to work with ActiveJob

### ISSUE THREAD

**Issue by xxx** [rwojsznis/sidekiq-lock#13]
Hi,

Thank you for making this gem.

I spent a little time getting it to work with ActiveJob (very small change at the end). I'd be happy to put up a PR for it, but I'm not sure how you would like to do the check. Right now my change exists as a small monkeypatch:

```ruby
module Sidekiq
  module Lock
    class Middleware
      def call(_worker, msg, _queue)
        klass = msg.fetch('wrapped').constantize
        options = klass.get_sidekiq_options['lock']

        setup_lock(options, msg['args']) unless options.nil?

        yield
      end
    end
  end
end
```

The issue is that the job class for ActiveJob is always a `ActiveJob::QueueAdapters::SidekiqAdapter::JobWrapper`, rather than the worker class directly, meaning that when it tried to `get_sidekiq_options['lock']`, they were unavailable. This patch merely gets to the correct class where the options actually exist.

The jobs themselves can be used as documented (make sure to `include Sidekiq::Lock::Worker` in the job class that needs locking functionality)

Is this behavior desirable? If so, how would you prefer that the conditional check be made?

**Comment by rwojsznis** [comment]
Hey there! Thank you for feedback and suggestion; this gem was never designed to be used with ActiveJob, but I will look into your approach and get back to you after next week+! (Currently traveling without a laptop)

**Comment by rwojsznis** [comment]
Sorry for the lag, got back this week!

Did some quick research:

- as per Sidekiq 6.0 you can use `sidekiq_options` directly; [reference](https://github.com/mperham/sidekiq/blob/master/Changes.md#60)

- seems we could [use this](https://github.com/mperham/sidekiq/blob/3330df0ee37cfd3e0cd3ef01e3e66b584b99d488/lib/sidekiq/worker.rb#L30) for obtaining options then

Hmm, so the _clean_ solution would be basically to require at least sidekiq 6.x and do some internal changes so we could support ActiveJob as well seamlessly 🤔 that's a pretty aggressive change tho - I will need to check what's the expected support for Sidekiq 5.x (like in terms of bugfixes) - I can only guess, but I think a lot of people could still use it

**Comment by xxx** [comment]
Yeah, we're using 6.whatever right now, so I'm just putting `sidekiq_options` in the job, as if it were a plain Sidekiq worker. That part works the exact same as Workers.

The only part that didn't work before was actually accessing those options from the middleware.

**Comment by rwojsznis** [comment]
Maybe we could just do a version check inside the middleware itself and use `sidekiq_options_hash` if version 6 is being used - simply for backward compatibility 🤔 such check shouldn't have too much of an performance impact (micro might be benchmark needed, also I totally forgot when middleware is being initialized - maybe we could do the version check once)

@xxx thanks for bringing that up in the first place! I will try to suggest some solution within this month (hopefully)

**Comment by yrashk** [comment]
I am also interested in exploring the ability to have this working with ActiveJob. Anything I can help with?

**Comment by rwojsznis** [comment]
@yrashk hello 👋 I'm kinda swamped with other work atm, if you have time to help that would be great! (the more contributors the merrier)

Let's maybe first prep failing test suite (so using lock inside ActveJob job, not sure what effort we're talking about here tbh 🙈 ) and take it from there - I think we quickly can agree on reasonable implementation that won't break backward compatibility, does that sounds good?

**Comment by rwojsznis** [comment]
Closing this due to lack of interest; maybe I will try to revisit it in next month or two but no promises 🙈
